src/Entities/Player.cpp

```diff
@@ -2643,6 +2643,64 @@ void cPlayer::SendBlocksAround(int a_BlockX, int a_BlockY, int a_BlockZ, int a_R
 
 bool cPlayer::PlaceBlocks(const sSetBlockVector & a_Blocks)
 {
+	// Check to see if any entity intersects any block being placed
+	class DoesIntersectBlock : public cEntityCallback
+	{
+	public:
+		const sSetBlockVector & m_Blocks;
+
+		// The distance inside the block the entity can still be.
+		const double EPSILON = 0.05;
+
+		DoesIntersectBlock(const sSetBlockVector & a_Blocks) :
+			m_Blocks(a_Blocks)
+		{
+		}
+
+		virtual bool Item(cEntity * a_Entity) override
+		{
+			cBoundingBox EntBox(a_Entity->GetPosition(), a_Entity->GetWidth() / 2, a_Entity->GetHeight());
+			for (auto blk: m_Blocks)
+			{
+				cBoundingBox BlockBox(
+					blk.GetX() + EPSILON,
+					blk.GetX() + 1.0 - EPSILON,
+					blk.GetY() + EPSILON,
+					blk.GetY() + 1.0 - EPSILON,
+					blk.GetZ() + EPSILON,
+					blk.GetZ() + 1.0 - EPSILON
+				);
+				if (EntBox.DoesIntersect(BlockBox))
+				{
+					return true;
+				}
+			}
+			return false;
+		}
+	} Callback(a_Blocks);
+
+	// Compute the bounding box of the blocks to be placed
+	double MaxDouble = std::numeric_limits<double>::max();
+	double MinDouble = std::numeric_limits<double>::min();
+	Vector3d BlocksMin(MaxDouble, MaxDouble, MaxDouble);
+	Vector3d BlocksMax(MinDouble, MinDouble, MinDouble);
+	for (auto blk: a_Blocks)
+	{
+		BlocksMin = BlocksMin.Min(blk.GetPos());
+		BlocksMax = BlocksMax.Max(blk.GetPos() + Vector3d(1, 1, 1));
+	}
```

Comment:
Stairs are half full-height and half half-height (a cube with an edge cut out).

Maybe the parameter could be the 6 cardinal neighbors (+/- X,Y,Z)? I don't think any block cares about blocks further away than that (except tripwire hooks, now that I research it).

At least with tripwire hooks, we could place it in a folded state, and when it immediately extends it just pushes the player. I'll have to load up a vanilla world later and see how that works.

For doors, both blocks are being placed at the same time, so asking "what's the bounding box if we place this block here" won't work, assuming we test one block at a time.

Which hints at another way to solve the problem: What if instead of "collision box" we had a different "placement box", which is fixed for a block type (or, block type and direction, to handle doors), and doesn't require neighbor information.

In vanilla, you can't place stairs if you're inside the full bounding cube (place a slab, sneak to the edge, place stairs on the ground next to the slab).